runtime/src/accounts_index.rs

```diff
@@ -1868,14 +1870,13 @@ impl<T: IndexValue> AccountsIndex<T> {
 
     pub fn add_root(&self, slot: Slot, caching_enabled: bool) {
         let mut w_roots_tracker = self.roots_tracker.write().unwrap();
+        // `AccountsDb::flush_accounts_cache()` relies on roots being added in order
+        assert!(slot >= w_roots_tracker.roots.max_inclusive());
```

Comment:
I don't fully understand all the context here, but I do see the assert has moved before the `insert()`. Does this have any impact on the greater-than-or-equal comparison? My initial thought would've expected this to change to strictly greater-than. Can you help me understand?

Reply:
inclusive is related and the fact it is an unsigned type is related. So, what is the initial value? The original code defaulated max_inclusive to 0. But, that clearly wasn't 'correct' because there are NO roots initially, so 0 (inclusive) is not a valid result for max_inclusive. So, the equal part of >= is here because of the 0 condition, I imagine, and probably then only for tests. The reason I had to move the assert before is because we now combine the state of 'max_inclusive' with the 'max_exclusive' maintained by roots tracker. So, once we insert 'slot' into roots tracker, we lose information on what the prior 'max_inclusive' USED to be. Since ALL of this happens in a write lock, it is legal and correct to check that slot is increasing before inserting the root in roots tracker.

Reply:
sorry - rollingbitfield is the thing that already has to keep track of max_exclusive to do its job.